### ISSUE TITLE
bge-reranker-base微调后产生的性能下降问题

### ISSUE THREAD

**Issue by Ziyang1060** [FlagOpen/FlagEmbedding#707]
项目发起者您好，我的领域数据集比较简单，几百条的样子，所以一开始用bge-reranker-base作初步尝试。
微调过程中logstep设置为1，能看到loss在稳定下降，所以我只训练了2个epoch，最终损失接近于0。

{'loss': 10.6336, 'grad_norm': 117.03962707519531, 'learning_rate': 9.9e-06, 'epoch': 0.2}                                                                        
{'loss': 8.5569, 'grad_norm': 107.7352066040039, 'learning_rate': 9.800000000000001e-06, 'epoch': 0.3}                                                            
{'loss': 6.7763, 'grad_norm': 96.62432098388672, 'learning_rate': 9.7e-06, 'epoch': 0.4}                                                                          
{'loss': 4.9323, 'grad_norm': 83.39602661132812, 'learning_rate': 9.600000000000001e-06, 'epoch': 0.5}                                                            
{'loss': 3.6468, 'grad_norm': 73.1478271484375, 'learning_rate': 9.5e-06, 'epoch': 0.6}                                                                           
{'loss': 2.4071, 'grad_norm': 60.45912551879883, 'learning_rate': 9.4e-06, 'epoch': 0.7}                                                                          
{'loss': 1.4583, 'grad_norm': 46.991539001464844, 'learning_rate': 9.3e-06, 'epoch': 0.8}                                                                         
{'loss': 0.9927, 'grad_norm': 35.67102813720703, 'learning_rate': 9.200000000000002e-06, 'epoch': 0.9}                                                            
{'loss': 0.4246, 'grad_norm': 20.708053588867188, 'learning_rate': 9.100000000000001e-06, 'epoch': 1.0}                                                           
{'loss': 0.182, 'grad_norm': 10.663714408874512, 'learning_rate': 9e-06, 'epoch': 1.1}                                                                            
{'loss': 0.0682, 'grad_norm': 4.576943874359131, 'learning_rate': 8.900000000000001e-06, 'epoch': 1.2}                                                            
{'loss': 0.0278, 'grad_norm': 2.014901876449585, 'learning_rate': 8.8e-06, 'epoch': 1.3}                                                                          
{'loss': 0.0152, 'grad_norm': 1.1478238105773926, 'learning_rate': 8.700000000000001e-06, 'epoch': 1.4}  

我的预期是，微调后模型在训练集上的评估指标（比如NDCG）要比没有微调前要好，但是实际上微调后评估指标反而差很多。
我观察了reranker的具体打分情况，微调前模型的打分基本都是正值，而微调后模型的打分都是负值了，而且会把我设置为负样本的items排的很靠前（这是评估指标变差的原因）。请问我之后应该如何进行修正？
    "bge-reranker-base": {
        "auc": 0.7775,
        "ndcg": 0.8723,
        "ndcg@1": 0.6908,
        "ndcg@5": 0.7278,
        "ndcg@10": 0.7044
    }
    "bge-reranker-base-ft-1": {
        "auc": 0.6457,
        "ndcg": 0.7048,
        "ndcg@1": 0.1596,
        "ndcg@5": 0.2526,
        "ndcg@10": 0.3063
    }

**Comment by Ziyang1060** [comment]
我补充一些我训练数据的设计策略: 
比如说对于一个query，我的排序结果为 [A>B>C>D]，并且 A B C 的相关度逐级递减
我会设置3个样本：
{query,pos:[A],neg:[B,C,D]}
{query,pos:[B],neg:[C,D]}
{query,pos:[C],neg:[D]}
请问我的这种设计策略会不会影响模型的训练过程呢？

**Comment by staoxiao** [comment]
请问一下您的训练参数以及评估数据是如何构造的？

**Comment by Ziyang1060** [comment]
感谢回复！
我的训练参数如下所示，我尝试调整过学习率和批量大小，将学习率调小后，微调过程中的loss变化更加缓慢，否则经过几个step就降为0了。
```
torchrun --nproc_per_node 2 \
-m FlagEmbedding.reranker.run \
--output_dir /data/zzy/Models/bge-reranker-base-ft-1 \
--model_name_or_path /data/zzy/Models/bge-reranker-base \
--train_data /home/zzy/Project/DrRank/data/reranker_training_data.jsonl \
--learning_rate 1e-5 \
--fp16 \
--num_train_epochs 2 \
--per_device_train_batch_size 4 \
--gradient_accumulation_steps 8 \
--dataloader_drop_last False \
--train_group_size 30 \
--max_len 512 \
--weight_decay 0.01 \
--logging_steps 1 \
--disable_tqdm False
```
为了简单起见，我的评估数据就是我的训练数据
训练数据：
{query,pos:[A],neg:[B,C,D]}
{query,pos:[B],neg:[C,D]}
{query,pos:[C],neg:[D]}
这三个样本对应了一条评估数据  query， [A>B>C>D]
具体来说会用reranker模型分别给A，B，C，D打分，按照打分进行降序排列，然后根据相关性等级计算NDCG指标。

**Comment by Ziyang1060** [comment]
作者您好，经过思考，我认为可能是当前的损失计算方式不太适合多级相关度的标签。我的训练数据设计策略也许会对训练造成恶劣影响，非常类似于https://github.com/FlagOpen/FlagEmbedding/issues/397提出的问题。

**Comment by staoxiao** [comment]
> 为了简单起见，我的评估数据就是我的训练数据
训练数据：
{query,pos:[A],neg:[B,C,D]}
{query,pos:[B],neg:[C,D]}
{query,pos:[C],neg:[D]}
这三个样本对应了一条评估数据 query， [A>B>C>D]
具体来说会用reranker模型分别给A，B，C，D打分，按照打分进行降序排列，然后根据相关性等级计算NDCG指标。

你的测试数据和训练数据完全一样吗？如果一样的话，loss这么小，评测指标应该不会这么低的。

另外，看到train_group_size很大，确认一下数据中确实有这么多neg，否则，会重复采样很多次，造成影响。

对于多级的训练，换一种更合适的损失应该要更好一些。如果你排在后面的C相关度已经不高，那么把他作为正样本会严重影响模型的训练。建议只对前几个相关的样本A，B进行训练。

损失函数可以换成与真实标签的KL loss。

**Comment by Ziyang1060** [comment]
感谢您的回复！
> 你的测试数据和训练数据完全一样吗？

我的测试数据和训练数据完全一样。训练过程的loss一切正常，但是在预测的时候反而会出现模型塌缩的情况。经过sigmoid函数规范化后，模型对于所有(query,document)的输出分数都趋近于0。

> 另外，看到train_group_size很大，确认一下数据中确实有这么多neg，否则，会重复采样很多次，造成影响。

我的评估数据 query， [A>....>B>....>C>....>D]，这个列表对应了几十条甚至一百多条的数据，只不过里面包括多级相关度（3，2，1，0，0代表不相关），所以我在构建训练数据集的时候，会把相关度等级为3/2/1的样本作为正样本，比正样本等级低的样本全作为负样本。如果这样的设计策略没问题的话，那么其负样本数量应该会大于我设置的train_group_size大小。

> 如果你排在后面的C相关度已经不高，那么把他作为正样本会严重影响模型的训练。建议只对前几个相关的样本A，B进行训练。

> 损失函数可以换成与真实标签的KL loss

感谢您的建议，我会尝试该策略。关于实现新的损失函数，我是否应该把项目克隆下来，然后修改损失函数代码后，通过`pip install -e .`命令安装呢？

**Comment by staoxiao** [comment]
> 感谢您的建议，我会尝试该策略。关于实现新的损失函数，我是否应该把项目克隆下来，然后修改损失函数代码后，通过pip install -e .命令安装呢？

对的
